File: scripts/memory/fact_pack.py
"""Fact Pack 解析 / 格式化（P1 注入用；P0 提供纯函数与自检）。"""

from __future__ import annotations


def parse_fact_pack(raw):
    """解析 Node 侧 /api/v2/memory/retrieve 返回的 Fact Pack。"""
    if not isinstance(raw, dict):
        return None
    facts = raw.get('facts')
    if not isinstance(facts, list):
        return None
    return {
        'facts': facts,
        'dropped': raw.get('dropped', []),
        'budget': raw.get('budget', {}),
    }


def format_fact_pack(facts):
    """把事实列表格式化为可注入任务文本的【记忆事实包】块。"""
    if not facts:
        return ''
    lines = ['【记忆事实包】']
    for f in facts[:50]:
        if not isinstance(f, dict):
            continue
        entity = str(f.get('entity') or '?')
        attribute = str(f.get('attribute') or 'value')
        value = str(f.get('value') or '')[:200]
        stance = str(f.get('stance') or 'neutral')
        source = str(f.get('source') or 'unknown')
        # 优先检索时的有效权重（存储权重 × 时间衰减 × 冲突惩罚）
        weight = f.get('effectiveWeight', f.get('weight'))
        lines.append(
            f'- #{f.get("id")} [{stance}/{source}] {entity}.{attribute} = {value} '
            f'(weight={weight})'
        )
    return '\n'.join(lines) + '\n'

File: src/memory/fact-pack.js
/**
 * Fact Pack 组装：按权重排序 + 预算裁剪。
 * 被裁剪的事实必须进入 dropped（不允许静默丢失，见方案 §5.4.3）。
 */

/**
 * 单条事实 → 可注入文本。
 * @param {object} [fact] Fact row (id / entity / attribute / value / source / stance / weight).
 * @returns {string} Formatted fact line for injection.
 */
export function formatFact(fact = {}) {
  const entity = String(fact.entity || '?');
  const attribute = String(fact.attribute || 'value');
  const value = String(fact.value ?? '').slice(0, 200);
  const source = String(fact.source || 'unknown');
  const stance = String(fact.stance || 'neutral');
  // 优先检索时的有效权重（存储权重 × 时间衰减 × 冲突惩罚）；缺失回退存储权重
  const weight = Number.isFinite(Number(fact.effectiveWeight))
    ? Number(fact.effectiveWeight)
    : (Number.isFinite(Number(fact.weight)) ? Number(fact.weight) : 0);
  return `- #${fact.id} [${stance}/${source}] ${entity}.${attribute} = ${value} (weight=${weight})`;
}

/**
 * 组装 Fact Pack。
 * @param {Array<object>} facts 已按 weight desc 排序
 * @param {object} opts Pack options.
 * @param {number} [opts.maxChars] Max total chars (default 2000).
 * @param {number} [opts.limit] Max item count (default 50).
 * @returns {{ facts: object[], dropped: Array<{ factId: number, entity: string, reason: string }>, budget: { used: number, max: number, limit: number } }} Picked facts + dropped reasons + budget usage.
 */
export function buildFactPack(facts, { maxChars = 2000, limit = 50 } = {}) {
  const picked = [];
  const dropped = [];
  let used = 0;
  const max = Math.max(200, Number(maxChars) || 2000);
  const maxItems = Math.max(1, Number(limit) || 50);

  for (const f of Array.isArray(facts) ? facts : []) {
    const line = formatFact(f);
    const size = line.length + 1;
    if (picked.length >= maxItems || used + size > max) {
      dropped.push({
        factId: f.id,
        entity: f.entity,
        reason: used + size > max ? 'budget' : 'limit',
      });
      continue;
    }
    picked.push(f);
    used += size;
  }

  return {
    facts: picked,
    dropped,
    budget: { used, max, limit: maxItems },
  };
}
